Sources/LanguageServerProtocol/AsyncQueue.swift

```diff
@@ -33,53 +33,97 @@ extension NSLock {
   }
 }
 
-/// A serial queue that allows the execution of asyncronous blocks of code.
+/// A queue that allows the execution of asyncronous blocks of code.
 public final class AsyncQueue {
-  /// Lock guarding `lastTask`.
-  private let lastTaskLock = NSLock()
+  public enum QueueKind {
+    /// A queue that allows concurrent execution of tasks.
+    case concurrent
 
-  /// The last scheduled task if it hasn't finished yet.
-  ///
-  /// Any newly scheduled tasks need to await this task to ensure that tasks are
-  /// executed syncronously.
-  ///
-  /// `id` is a unique value to identify the task. This allows us to set `lastTask`
-  /// to `nil` if the queue runs empty.
-  private var lastTask: (task: AnyTask, id: UUID)?
+    /// A queue that executes one task after the other.
+    case serial
+  }
+
+  private struct PendingTask {
+    /// The task that is pending.
+    let task: any AnyTask
+
+    /// Whether the task needs to finish executing befoer any other task can
+    /// start in executing in the queue.
+    let isBarrier: Bool
+
+    /// A unique value used to identify the task. This allows tasks to get
+    /// removed from `pendingTasks` again after they finished executing.
+    let id: UUID
+  }
+
+  /// Whether the queue allows concurrent execution of tasks.
+  private let kind: QueueKind
+
+  ///  Lock guarding `pendingTasks`.
+  private let pendingTasksLock = NSLock()
 
-  public init() {
-    self.lastTaskLock.name = "AsyncQueue.lastTaskLock"
+  /// Pending tasks that have not finished execution yet.
+  private var pendingTasks = [PendingTask]()
+
+  public init(_ kind: QueueKind) {
+    self.kind = kind
+    self.pendingTasksLock.name = "AsyncQueue"
   }
 
   /// Schedule a new closure to be executed on the queue.
   ///
-  /// All previously added tasks are guaranteed to finished executing before
-  /// this closure gets executed.
+  /// If this is a serial queue, all previously added tasks are guaranteed to
+  /// finished executing before this closure gets executed.
+  ///
+  /// If this is a barrier, all previously scheduled tasks are guaranteed to
+  /// finish execution before the barrier is executed and all tasks that are
+  /// added later will wait until the barrier finishes execution.
   @discardableResult
   public func async<Success: Sendable>(
     priority: TaskPriority? = nil,
+    barrier isBarrier: Bool = false,
     @_inheritActorContext operation: @escaping @Sendable () async -> Success
   ) -> Task<Success, Never> {
+    // We treat a serial queue just like a concurrent queue in which every task
+    // is a barrier.
+    let isBarrier = isBarrier || kind == .serial
```

Comment:
Is this a holdover from a previous design? Best I can tell, when the kind is `.serial`, we don't care about `isBarrier`

Reply:
No, this is essential and is where the serial part of the queue comes in. The queue itself doesn’t really have an inherrent notion of being serial or concurrent (we don’t use dispatch queues or anything like that). It’s just that if the queue is serial, every task gets enqueued as a barrier, which means that they will be executed one after the other.

Reply:
Hmm, wait. Actually that’s not true and you are right, we don’t need this. Forget my comment above.